### PR TITLE
Android kotlin

### PR DESCRIPTION
**Pull request by brunokrebs**

### CODE REVIEW

File: _posts/2017-06-20-authenticating-android-apps-developed-with-kotlin.markdown [hunk 1]
```diff
@@ -0,0 +1,632 @@
...
+- The `getHeaders` function to mark the `Content-Type` of the request as `text/plain`, and to add the `access_token` of the authenticated user.
+
+These changes are everything that we need in our Kotlin app to properly communicate with the secured RESTful endpoint. Now, whenever a user types an item in the text input, our Kotlin app sends a request to the backend with the item and the `access_token`. The backend then gets this `access_token`, validates it [against a public key provided by Auth0](https://auth0.com/blog/navigating-rs256-and-jwks/) and then adds the item received to the to do list.
+
+If we run our application now, we will be able to use all its features. We will be able to see the public (read-only) list of to do items and then, if we authenticate ourselves, we will be able to update this list.
+
+## Building a Home Made Security Solution
+
+In case we don't want to rely on state-of-the-art features provided by Auth0, we can also develop our home made security solution with JWTs. We won't get into the nitty-gritty details of the refactoring process, but here it is an overview of the steps needed to grown our own solution:
```
Comment (unicodeveloper) [inline _posts/2017-06-20-authenticating-android-apps-developed-with-kotlin.markdown]: `needed to grown our own solution`

=>

`needed to grow our own solution`

File: _posts/2017-06-20-authenticating-android-apps-developed-with-kotlin.markdown [hunk 2]
```diff
@@ -0,0 +1,632 @@
...
+
+- [Securing Node.js with JWTs](https://auth0.com/blog/building-and-authenticating-nodejs-apps/)
+- [Securing Spring Boot with JWTs](https://auth0.com/blog/securing-spring-boot-with-jwts/)
+- [Securing .NET Core with JWTs](https://auth0.com/blog/asp-dot-net-core-authentication-tutorial/)
+
+You can check these resources if you need to develop your own solution.
+
+## Conclusion
+
+Developing a secure Android application with Kotlin is trivial, as we managed to see in this blog post. If you had no previous Kotlin knowledge, chances are that you hadn't had trouble to follow along. Kotlin integration with existing Java libraries and frameworks make the process of learning it smooth. But, to use this language to its full power, a good amount of coding hours and studying is need.
```
Comment (unicodeveloper) [inline _posts/2017-06-20-authenticating-android-apps-developed-with-kotlin.markdown]: I would rephrase this from

`chances are that you hadn't had trouble to follow along`

to

`chances are that you had no trouble following along`

File: _posts/2017-06-20-authenticating-android-apps-developed-with-kotlin.markdown [hunk 3]
```diff
@@ -0,0 +1,632 @@
...
+author:
+  name: "Bruno Krebs"
+  url: "https://twitter.com/brunoskrebs"
+  mail: "bruno.krebs@auth0.com"
+  avatar: "https://www.gravatar.com/avatar/76ea40cbf67675babe924eecf167b9b8?s=60"
+design:
+  bg_color: "#7370B3"
+  image: https://cdn.auth0.com/blog/create-kotlin-app/logo.png
+tags:
+- android
```
Comment (unicodeveloper) [inline _posts/2017-06-20-authenticating-android-apps-developed-with-kotlin.markdown]: can we have a java tag here too?

File: _posts/2017-06-20-authenticating-android-apps-developed-with-kotlin.markdown [hunk 4]
```diff
@@ -0,0 +1,632 @@
...
+
+- [Kotlin Reference](https://kotlinlang.org/docs/reference/)—where the details of Kotlin's syntax are explained
+- [Try Kotlin](https://try.kotlinlang.org)—where you can do some hands on exercises to learn Kotlin.
+- [Kotlin in Action](https://manning.com/books/kotlin-in-action)—if you want to dive deep into this new language
+
+Besides that, be aware that JetBrains has been investing a lot on tools to help Java developers migrate to Kotlin. For example, IntelliJ and Android Studio 3.0 are both shipped with tools that automatically translates Java source code to Kotlin.
+
+## Developing an Android App in Kotlin
+
+To start a new project, you can take advantage of the *Create New Project* wizard that comes with Android Studio. Its usage is quite simple, and just a few steps are needed, [as shown in this tutorial](https://kotlinlang.org/docs/tutorials/kotlin-android.html). But, to speed things up, we are going to [clone this repository](https://github.com/auth0-blog/kotlin-app) in our local machine.
```
Comment (brunokrebs) [inline _posts/2017-06-20-authenticating-android-apps-developed-with-kotlin.markdown]: Added a note below

File: _posts/2017-06-20-authenticating-android-apps-developed-with-kotlin.markdown (hunk 3 above)
Comment (brunokrebs) [inline _posts/2017-06-20-authenticating-android-apps-developed-with-kotlin.markdown]: no need, I think

File: _posts/2017-06-20-authenticating-android-apps-developed-with-kotlin.markdown [hunk 5]
```diff
@@ -0,0 +1,632 @@
...
+
+    override fun onCreate(savedInstanceState: Bundle?) {
+        super.onCreate(savedInstanceState)
+        setContentView(R.layout.activity_main)
+
+        // setting up a Volley RequestQueue
+        val queue = Volley.newRequestQueue(this)
+
+        // getting a reference for the ListView
+        val listToDo = findViewById(R.id.list_todo) as ListView
```
Comment (brunokrebs) [inline _posts/2017-06-20-authenticating-android-apps-developed-with-kotlin.markdown]: I understood but decided to leave untouched as this is not the focus of this article

File: _posts/2017-06-20-authenticating-android-apps-developed-with-kotlin.markdown [hunk 6]
```diff
@@ -0,0 +1,632 @@
...
+            android:layout_width="match_parent"
+            android:layout_height="wrap_content"
+            android:text="Add item"
+            android:visibility="@{loggedIn ? View.VISIBLE : View.GONE}" />
+        <!-- ... -->
+    </LinearLayout>
+</layout>
+```
+
+Both elements are going to be shown only if the user is `loggedIn`, as defined in their `android:visibility` properties. We can now add a listener, that will trigger a POST request to the backend, to the new button. We achieve that by changing the `MainActivity` as follows:
```
Comment (brunokrebs) [inline _posts/2017-06-20-authenticating-android-apps-developed-with-kotlin.markdown]: I don't see why change this.

File: _posts/2017-06-20-authenticating-android-apps-developed-with-kotlin.markdown [hunk 7]
```diff
@@ -415,28 +353,91 @@ class MainActivity : AppCompatActivity() {
...
+
+    fun saveCredentials(context: Context, credentials: Credentials) {
+        val sp = context.getSharedPreferences(
+                PREFERENCES_NAME, Context.MODE_PRIVATE)
+
+        sp!!.edit().putString(ACCESS_TOKEN, credentials.accessToken)
+                .apply()
+    }
+
+    fun getAccessToken(context: Context): String {
```
Comment (lbalmaceda) [inline _posts/2017-06-20-authenticating-android-apps-developed-with-kotlin.markdown]: This one is named "getAccessToken" and it returns the access_token, but the other one is named "saveCredentials" and it's just saving the access_token. I'd name it `saveAccessToken` or `setAccessToken`. (change it in the rest of the article where it's used as well.)

File: _posts/2017-06-20-authenticating-android-apps-developed-with-kotlin.markdown [hunk 8]
```diff
@@ -362,7 +297,7 @@ After that, we will modify the `MainActivity` and its layout to add a button tha
 </layout>
 ```
 
-> We would usually define labels and static texts in the `strings.xml` file. However, to simplify the tutorial, we will leave it defined in the layout file.
+> We would usually define labels and static texts in the `strings.xml` file. However, to simplify the tutorial, we will leave them defined in the layout file.
```
Comment (lbalmaceda) [inline _posts/2017-06-20-authenticating-android-apps-developed-with-kotlin.markdown]: You actually have a section below that explains how to add texts to the `strings.xml` file, so this line is lying 😄
